src/NSubstitute/Core/Arguments/NonParamsArgumentSpecificationFactory.cs

```diff
@@ -24,7 +24,7 @@ public IArgumentSpecification Create(object argument, IParameterInfo parameterIn
                 return _argumentEqualsSpecificationFactory.Create(argument, parameterInfo.ParameterType);
             }
 
-            throw new AmbiguousArgumentsException(suppliedArgumentSpecifications.AllSpecifications);
+            throw new AmbiguousArgumentsException();
```

Comment:
As per previous comment, I'd prefer to create these with the required information if possible.
I've always been lost in this area of the code though, so happy to go with whatever you decide is best here. 😂 (If it's quick to explain it would be good to get a quick explanation of why you've dropped the `AllSpecifications` argument off here, but if it is going to take more than a couple of minutes to type up then it's not worth it, your time is valuable so don't waste it :) )

Reply:
Sure, initially I also wanted to provide all the data here as it look nicer. However, currently exception needs more data that is available in this layer. We have the following sequence:
```
MixedArgumentSpecificationsFactory
    --> ArgumentSpecificationFactory
        --> NonParamsArgumentSpecificationFactory
```

The `MixedArgumentSpecificationsFactory` iterates over each passed argument and asks the underlying levels to resolve a matcher for that argument.
In this PR I extend the raised exception so it should know the already resolved matchers for the better diagnostics. The `NonParamsArgumentSpecificationFactory` knows only about the current argument, so it cannot throw the final exception with all the data. To mitigate that I capture the `AmbiguousArgumentException` in the `MixedArgumentSpecificationsFactory` class and re-throw it with the extended information I have at this level.

With such a design the underlying exception can be as dummy as possible - we intercept it in any case and throw a new one. That's why I create it very empty and pass all the specification at the upper layer. That allowed to remove the `AllSpecifications` API as it's no longer required.

Of course, there might be an alternative approach when we pass info about the already resolved specs to the underlying levels, but it looks like a poor design for me.